In [1]:
%reset

Once deleted, variables cannot be recovered. Proceed (y/[n])?  y


In [2]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns

In [3]:
from statsmodels.formula.api import ols
import statsmodels.api as sm

In [14]:
# Single Replicate of the 2^4 design
A=pd.Series([-1,1,-1,1,-1,+1,-1,+1,-1,1,-1,1,-1,1,-1,1],name="Temp")
B=pd.Series([-1,-1,1,1,-1,-1,1,1,-1,-1,1,1,-1,-1,1,1], name='Pressure')
C=pd.Series([-1,-1,-1,-1,1,1,1,1,-1,-1,-1,-1,1,1,1,1],name='Conc')
D=pd.Series([-1,-1,-1,-1,-1,-1,-1,-1,1,1,1,1,1,1,1,1],name="Stirring")
OC=pd.Series([45,71,48,65,68,60,80,65,43,100,45,104,75,86,70,96],name="Filtration")

In [9]:
DFA=pd.concat([A,B,C,D,OC],axis='columns')
print(DFA)

    Temp  Pressure  Conc  Stirring  Filtration
0     -1        -1    -1        -1          45
1      1        -1    -1        -1          71
2     -1         1    -1        -1          48
3      1         1    -1        -1          65
4     -1        -1     1        -1          68
5      1        -1     1        -1          60
6     -1         1     1        -1          80
7      1         1     1        -1          65
8     -1        -1    -1         1          43
9      1        -1    -1         1         100
10    -1         1    -1         1          45
11     1         1    -1         1         104
12    -1        -1     1         1          75
13     1        -1     1         1          86
14    -1         1     1         1          70
15     1         1     1         1          96


In [10]:
from statsmodels.formula.api import ols
import statsmodels.api as sm

model1=ols('OC~A*B*C*D',data=DFA)
results1=model1.fit()
print(results1.summary2())

                 Results: Ordinary least squares
Model:              OLS              Adj. R-squared:     nan      
Dependent Variable: OC               AIC:                -907.9467
Date:               2022-11-30 11:03 BIC:                -895.5852
No. Observations:   16               Log-Likelihood:     469.97   
Df Model:           15               F-statistic:        nan      
Df Residuals:       0                Prob (F-statistic): nan      
R-squared:          1.000            Scale:              inf      
--------------------------------------------------------------------
               Coef.    Std.Err.      t      P>|t|   [0.025   0.975]
--------------------------------------------------------------------
Intercept     70.0625        inf    0.0000     nan      nan      nan
A             10.8125        inf    0.0000     nan      nan      nan
B              1.5625        inf    0.0000     nan      nan      nan
A:B            0.0625        inf    0.0000     nan      nan      nan

C:\Users\carol\anaconda3\lib\site-packages\scipy\stats\_stats_py.py:1769: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=16
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "
C:\Users\carol\anaconda3\lib\site-packages\statsmodels\regression\linear_model.py:1671: RuntimeWarning: divide by zero encountered in double_scalars
  return np.dot(wresid, wresid) / self.df_resid
C:\Users\carol\anaconda3\lib\site-packages\statsmodels\base\model.py:1512: RuntimeWarning: invalid value encountered in multiply
  cov_p = self.normalized_cov_params * scale
C:\Users\carol\anaconda3\lib\site-packages\statsmodels\regression\linear_model.py:1749: RuntimeWarning: divide by zero encountered in true_divide
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
C:\Users\carol\anaconda3\lib\site-packages\statsmodels\regression\linear_model.py:1749: RuntimeWarning: invalid value encountered in double_scalars
  return 1 - (np.divide(self.nobs - self.k_constan

In [11]:
from statsmodels.stats.anova import anova_lm
anova_results1 = anova_lm(results1)
print(anova_results1)

           df        sum_sq    mean_sq    F  PR(>F)
A         1.0  1.870562e+03  1870.5625  0.0     NaN
B         1.0  3.906250e+01    39.0625  0.0     NaN
A:B       1.0  6.250000e-02     0.0625  0.0     NaN
C         1.0  3.900625e+02   390.0625  0.0     NaN
A:C       1.0  1.314062e+03  1314.0625  0.0     NaN
B:C       1.0  2.256250e+01    22.5625  0.0     NaN
A:B:C     1.0  1.406250e+01    14.0625  0.0     NaN
D         1.0  8.555625e+02   855.5625  0.0     NaN
A:D       1.0  1.105562e+03  1105.5625  0.0     NaN
B:D       1.0  5.625000e-01     0.5625  0.0     NaN
A:B:D     1.0  6.806250e+01    68.0625  0.0     NaN
C:D       1.0  5.062500e+00     5.0625  0.0     NaN
A:C:D     1.0  1.056250e+01    10.5625  0.0     NaN
B:C:D     1.0  2.756250e+01    27.5625  0.0     NaN
A:B:C:D   1.0  7.562500e+00     7.5625  0.0     NaN
Residual  0.0  2.872716e-26        inf  NaN     NaN


C:\Users\carol\anaconda3\lib\site-packages\statsmodels\stats\anova.py:138: RuntimeWarning: divide by zero encountered in double_scalars
  (model.ssr / model.df_resid))


In [15]:
# Refit the model with A, C D
model2=ols('OC~A*C*D',data=DFA)
results2=model2.fit()
print(results2.summary2())

                 Results: Ordinary least squares
Model:              OLS              Adj. R-squared:     0.941   
Dependent Variable: OC               AIC:                100.0874
Date:               2022-11-30 11:18 BIC:                106.2681
No. Observations:   16               Log-Likelihood:     -42.044 
Df Model:           7                F-statistic:        35.35   
Df Residuals:       8                Prob (F-statistic): 2.12e-05
R-squared:          0.969            Scale:              22.438  
------------------------------------------------------------------
              Coef.   Std.Err.     t     P>|t|    [0.025    0.975]
------------------------------------------------------------------
Intercept    70.0625    1.1842  59.1641  0.0000   67.3317  72.7933
A            10.8125    1.1842   9.1306  0.0000    8.0817  13.5433
C             4.9375    1.1842   4.1695  0.0031    2.2067   7.6683
A:C          -9.0625    1.1842  -7.6528  0.0001  -11.7933  -6.3317
D             7.3125

C:\Users\carol\anaconda3\lib\site-packages\scipy\stats\_stats_py.py:1769: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=16
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


In [17]:
anova_results2 = anova_lm(results2)
print(anova_results2)

           df     sum_sq    mean_sq          F    PR(>F)
A         1.0  1870.5625  1870.5625  83.367688  0.000017
C         1.0   390.0625   390.0625  17.384401  0.003124
A:C       1.0  1314.0625  1314.0625  58.565460  0.000060
D         1.0   855.5625   855.5625  38.130919  0.000267
A:D       1.0  1105.5625  1105.5625  49.272981  0.000110
C:D       1.0     5.0625     5.0625   0.225627  0.647483
A:C:D     1.0    10.5625    10.5625   0.470752  0.512032
Residual  8.0   179.5000    22.4375        NaN       NaN


# Randomized Block Design
Ex. Confounding the 2^4 factorial design in 2 blocks

Introduce a block effect where the raw material in one block (block A = +1) is of poorer quality and therefor results in a 20 unit reduction in coversion yield.  

In [18]:
A=pd.Series([-1,1,-1,1,-1,+1,-1,+1,-1,1,-1,1,-1,1,-1,1],name="Temp")
B=pd.Series([-1,-1,1,1,-1,-1,1,1,-1,-1,1,1,-1,-1,1,1], name='Pressure')
C=pd.Series([-1,-1,-1,-1,1,1,1,1,-1,-1,-1,-1,1,1,1,1],name='Conc')
D=pd.Series([-1,-1,-1,-1,-1,-1,-1,-1,1,1,1,1,1,1,1,1],name="Stirring")
E=pd.Series([1,-1,-1,1,-1,1,1,-1,-1,1,1,-1,1,-1,-1,1],name='Block')
OC=pd.Series([25,71,48,45,68,40,60,65,43,80,25,104,55,86,70,76],name="Filtration")

In [19]:
DF=pd.concat([A,B,C,D,E,OC],axis='columns')
print(DF)

    Temp  Pressure  Conc  Stirring  Block  Filtration
0     -1        -1    -1        -1      1          25
1      1        -1    -1        -1     -1          71
2     -1         1    -1        -1     -1          48
3      1         1    -1        -1      1          45
4     -1        -1     1        -1     -1          68
5      1        -1     1        -1      1          40
6     -1         1     1        -1      1          60
7      1         1     1        -1     -1          65
8     -1        -1    -1         1     -1          43
9      1        -1    -1         1      1          80
10    -1         1    -1         1      1          25
11     1         1    -1         1     -1         104
12    -1        -1     1         1      1          55
13     1        -1     1         1     -1          86
14    -1         1     1         1     -1          70
15     1         1     1         1      1          76


In [23]:
from statsmodels.formula.api import ols
import statsmodels.api as sm

model3=ols('OC~Temp+Pressure+Conc+Stirring+Block+Temp*Pressure+Temp*Conc+Temp*Stirring+Pressure*Conc+Pressure*Stirring+Conc*Stirring+Temp*Pressure*Conc+Pressure*Conc*Stirring+Temp*Conc*Stirring+Temp*Pressure*Stirring',data=DF)
results3=model3.fit()
print(results3.summary2())

                  Results: Ordinary least squares
Model:               OLS              Adj. R-squared:     nan      
Dependent Variable:  OC               AIC:                -908.8430
Date:                2022-11-30 11:19 BIC:                -896.4816
No. Observations:    16               Log-Likelihood:     470.42   
Df Model:            15               F-statistic:        nan      
Df Residuals:        0                Prob (F-statistic): nan      
R-squared:           1.000            Scale:              inf      
-------------------------------------------------------------------
                        Coef.  Std.Err.    t    P>|t| [0.025 0.975]
-------------------------------------------------------------------
Intercept              60.0625      inf  0.0000   nan    nan    nan
Temp                   10.8125      inf  0.0000   nan    nan    nan
Pressure                1.5625      inf  0.0000   nan    nan    nan
Conc                    4.9375      inf  0.0000   nan    nan    na

C:\Users\carol\anaconda3\lib\site-packages\scipy\stats\_stats_py.py:1769: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=16
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "
C:\Users\carol\anaconda3\lib\site-packages\statsmodels\regression\linear_model.py:1671: RuntimeWarning: divide by zero encountered in double_scalars
  return np.dot(wresid, wresid) / self.df_resid
C:\Users\carol\anaconda3\lib\site-packages\statsmodels\regression\linear_model.py:1749: RuntimeWarning: divide by zero encountered in true_divide
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
C:\Users\carol\anaconda3\lib\site-packages\statsmodels\regression\linear_model.py:1749: RuntimeWarning: invalid value encountered in double_scalars
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)


In [24]:
from statsmodels.stats.anova import anova_lm
anova_results3 = anova_lm(results3)
print(anova_results3)

                         df        sum_sq    mean_sq    F  PR(>F)
Temp                    1.0  1.870562e+03  1870.5625  0.0     NaN
Pressure                1.0  3.906250e+01    39.0625  0.0     NaN
Conc                    1.0  3.900625e+02   390.0625  0.0     NaN
Stirring                1.0  8.555625e+02   855.5625  0.0     NaN
Block                   1.0  1.387562e+03  1387.5625  0.0     NaN
Temp:Pressure           1.0  6.250000e-02     0.0625  0.0     NaN
Temp:Conc               1.0  1.314062e+03  1314.0625  0.0     NaN
Temp:Stirring           1.0  1.105563e+03  1105.5625  0.0     NaN
Pressure:Conc           1.0  2.256250e+01    22.5625  0.0     NaN
Pressure:Stirring       1.0  5.625000e-01     0.5625  0.0     NaN
Conc:Stirring           1.0  5.062500e+00     5.0625  0.0     NaN
Temp:Pressure:Conc      1.0  1.406250e+01    14.0625  0.0     NaN
Pressure:Conc:Stirring  1.0  2.756250e+01    27.5625  0.0     NaN
Temp:Conc:Stirring      1.0  1.056250e+01    10.5625  0.0     NaN
Temp:Press

C:\Users\carol\anaconda3\lib\site-packages\statsmodels\stats\anova.py:138: RuntimeWarning: divide by zero encountered in double_scalars
  (model.ssr / model.df_resid))


Refit the model with meaningful effects

In [27]:
model4=ols('OC~Temp+Conc+Stirring+Temp*Conc+Temp*Stirring+Block',data=DF)
results4=model4.fit()
print(results4.summary2())

                 Results: Ordinary least squares
Model:              OLS              Adj. R-squared:     0.956   
Dependent Variable: OC               AIC:                98.7904 
Date:               2022-11-30 11:31 BIC:                104.1985
No. Observations:   16               Log-Likelihood:     -42.395 
Df Model:           6                F-statistic:        55.37   
Df Residuals:       9                Prob (F-statistic): 1.35e-06
R-squared:          0.974            Scale:              20.840  
-----------------------------------------------------------------
                  Coef.  Std.Err.    t    P>|t|   [0.025   0.975]
-----------------------------------------------------------------
Intercept        60.0625   1.1413 52.6274 0.0000  57.4807 62.6443
Temp             10.8125   1.1413  9.4740 0.0000   8.2307 13.3943
Conc              4.9375   1.1413  4.3263 0.0019   2.3557  7.5193
Stirring          7.3125   1.1413  6.4073 0.0001   4.7307  9.8943
Temp:Conc        -9.0625   

C:\Users\carol\anaconda3\lib\site-packages\scipy\stats\_stats_py.py:1769: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=16
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


In [28]:
anova_results4 = anova_lm(results4)
print(anova_results4)

                df     sum_sq      mean_sq          F    PR(>F)
Temp           1.0  1870.5625  1870.562500  89.757081  0.000006
Conc           1.0   390.0625   390.062500  18.716761  0.001915
Stirring       1.0   855.5625   855.562500  41.053316  0.000124
Temp:Conc      1.0  1314.0625  1314.062500  63.053982  0.000023
Temp:Stirring  1.0  1105.5625  1105.562500  53.049317  0.000046
Block          1.0  1387.5625  1387.562500  66.580806  0.000019
Residual       9.0   187.5625    20.840278        NaN       NaN


The Block effect now accumlates and conflates the block effect and the 4-way ABCD interaction.  The ABCD interaction previously (without the block effect) was 0.6875 for going from 0 to 1 in the ols model (so double this to 1.375 from -1 to 1).  

Now, with the block effect, we have the block effect plus the ABCD interaction = -9.3215 (so double this to -18.643).  The block effect is -18.643 - 1.375 = -20.  This is the reduction we put in to show the poorer quality raw material used in block A = 1.  

This is a noise reductuion.  The added variability of the block gets distributed across other design factors and leads to lost ability to find true conclusions.  If you run the unblocked version of this experiment without accounting for the block, you get a high effect size for the interaction.  You have to include everything in the model and cannot reduce and simplify the model.  

In [29]:
model5=ols('OC~Temp*Pressure*Conc*Stirring',data=DF)
results5=model5.fit()
print(results5.summary2())

                    Results: Ordinary least squares
Model:                OLS                Adj. R-squared:       nan      
Dependent Variable:   OC                 AIC:                  -909.8558
Date:                 2022-11-30 11:51   BIC:                  -897.4944
No. Observations:     16                 Log-Likelihood:       470.93   
Df Model:             15                 F-statistic:          nan      
Df Residuals:         0                  Prob (F-statistic):   nan      
R-squared:            1.000              Scale:                inf      
------------------------------------------------------------------------
                             Coef.  Std.Err.    t    P>|t| [0.025 0.975]
------------------------------------------------------------------------
Intercept                   60.0625      inf  0.0000   nan    nan    nan
Temp                        10.8125      inf  0.0000   nan    nan    nan
Pressure                     1.5625      inf  0.0000   nan    nan    nan

C:\Users\carol\anaconda3\lib\site-packages\scipy\stats\_stats_py.py:1769: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=16
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "
C:\Users\carol\anaconda3\lib\site-packages\statsmodels\regression\linear_model.py:1671: RuntimeWarning: divide by zero encountered in double_scalars
  return np.dot(wresid, wresid) / self.df_resid
C:\Users\carol\anaconda3\lib\site-packages\statsmodels\base\model.py:1512: RuntimeWarning: invalid value encountered in multiply
  cov_p = self.normalized_cov_params * scale
C:\Users\carol\anaconda3\lib\site-packages\statsmodels\regression\linear_model.py:1749: RuntimeWarning: divide by zero encountered in true_divide
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
C:\Users\carol\anaconda3\lib\site-packages\statsmodels\regression\linear_model.py:1749: RuntimeWarning: invalid value encountered in double_scalars
  return 1 - (np.divide(self.nobs - self.k_constan

If you account the 4-way interaction as noise and refit the model with Temp, Conc, and Stirring:

In [33]:
model6=ols('OC~Temp+Pressure+Conc+Stirring+Temp*Conc+Temp*Stirring',data=DF)
results6=model6.fit()
print(results6.summary2())

                 Results: Ordinary least squares
Model:              OLS              Adj. R-squared:     0.640   
Dependent Variable: OC               AIC:                132.4363
Date:               2022-11-30 12:01 BIC:                137.8444
No. Observations:   16               Log-Likelihood:     -59.218 
Df Model:           6                F-statistic:        5.444   
Df Residuals:       9                Prob (F-statistic): 0.0123  
R-squared:          0.784            Scale:              170.67  
-----------------------------------------------------------------
                  Coef.  Std.Err.    t    P>|t|   [0.025   0.975]
-----------------------------------------------------------------
Intercept        60.0625   3.2661 18.3899 0.0000  52.6742 67.4508
Temp             10.8125   3.2661  3.3106 0.0091   3.4242 18.2008
Pressure          1.5625   3.2661  0.4784 0.6438  -5.8258  8.9508
Conc              4.9375   3.2661  1.5118 0.1649  -2.4508 12.3258
Stirring          7.3125   

C:\Users\carol\anaconda3\lib\site-packages\scipy\stats\_stats_py.py:1769: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=16
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


In [34]:
anova_results6 = anova_lm(results6)
print(anova_results6)

                df     sum_sq      mean_sq          F    PR(>F)
Temp           1.0  1870.5625  1870.562500  10.959881  0.009076
Pressure       1.0    39.0625    39.062500   0.228873  0.643769
Conc           1.0   390.0625   390.062500   2.285429  0.164882
Stirring       1.0   855.5625   855.562500   5.012858  0.051933
Temp:Conc      1.0  1314.0625  1314.062500   7.699272  0.021585
Temp:Stirring  1.0  1105.5625  1105.562500   6.477642  0.031445
Residual       9.0  1536.0625   170.673611        NaN       NaN


Then the MSE for the model is 170.67.  This is higher than the one found with the blocking effect, where the MSE for that model is 20.84.  